url86119f019a_solzy_hive_parse_tgz-pJ8-org_apache_hadoop_hive_ql_parse_MainJ8/src/org/apache/hadoop/hive/ql/parse/HiveParser_IdentifiersParser.java

```diff
@@ -0,0 +1,12423 @@
+							alt20=1;
+						}
+
+						switch (alt20) {
+						case 1 :
+							// IdentifiersParser.g:162:7: ( COMMA )=> COMMA columnRefOrder
+							{
+							COMMA84=(Token)match(input,COMMA,FOLLOW_COMMA_in_sortByClause931); if (state.failed) return retval; 
+							if ( state.backtracking==0 ) stream_COMMA.add(COMMA84);
+
+							pushFollow(FOLLOW_columnRefOrder_in_sortByClause933);
+							columnRefOrder85=gHiveParser.columnRefOrder();
+							state._fsp--;
+							if (state.failed) return retval;
+							if ( state.backtracking==0 ) stream_columnRefOrder.add(columnRefOrder85.getTree());
+							}
+							break;
+
+						default :
+							break loop20;
+						}
+					}
+
+					// AST REWRITE
+					// elements: columnRefOrder
+					// token labels: 
+					// rule labels: retval
+					// token list labels: 
+					// rule list labels: 
+					// wildcard labels: 
+					if ( state.backtracking==0 ) {
+					retval.tree = root_0;
+					RewriteRuleSubtreeStream stream_retval=new RewriteRuleSubtreeStream(adaptor,"rule retval",retval!=null?retval.getTree():null);
+
+					root_0 = (CommonTree)adaptor.nil();
+					// 162:40: -> ^( TOK_SORTBY ( columnRefOrder )+ )
+					{
+						// IdentifiersParser.g:162:43: ^( TOK_SORTBY ( columnRefOrder )+ )
+						{
+						CommonTree root_1 = (CommonTree)adaptor.nil();
+						root_1 = (CommonTree)adaptor.becomeRoot((CommonTree)adaptor.create(TOK_SORTBY, "TOK_SORTBY"), root_1);
+						if ( !(stream_columnRefOrder.hasNext()) ) {
+							throw new RewriteEarlyExitException();
+						}
+						while ( stream_columnRefOrder.hasNext() ) {
+							adaptor.addChild(root_1, stream_columnRefOrder.nextTree());
+						}
+						stream_columnRefOrder.reset();
+
+						adaptor.addChild(root_0, root_1);
+						}
+
+					}
+
+
+					retval.tree = root_0;
+					}
+
+					}
+					break;
+
+			}
+			retval.stop = input.LT(-1);
+
+			if ( state.backtracking==0 ) {
+			retval.tree = (CommonTree)adaptor.rulePostProcessing(root_0);
+			adaptor.setTokenBoundaries(retval.tree, retval.start, retval.stop);
+			}
+			if ( state.backtracking==0 ) { gParent.msgs.pop(); }
+		}
+
+		catch (RecognitionException e) {
+		  throw e;
+		}
+
+		finally {
+			// do for sure before leaving
+		}
+		return retval;
+	}
+	// $ANTLR end "sortByClause"
+
+
+	public static class function_return extends ParserRuleReturnScope {
+		CommonTree tree;
+		@Override
+		public CommonTree getTree() { return tree; }
+	};
+
+
+	// $ANTLR start "function"
+	// IdentifiersParser.g:166:1: function : functionName LPAREN ( (star= STAR ) | (dist= KW_DISTINCT )? ( selectExpression ( COMMA selectExpression )* )? ) RPAREN ( KW_OVER ws= window_specification )? -> {$star != null}? ^( TOK_FUNCTIONSTAR functionName ( $ws)? ) -> {$dist == null}? ^( TOK_FUNCTION functionName ( ( selectExpression )+ )? ( $ws)? ) -> ^( TOK_FUNCTIONDI functionName ( ( selectExpression )+ )? ) ;
+	public final HiveParser_IdentifiersParser.function_return function() throws RecognitionException {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 209. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.